### PR TITLE
Create new QuestionForm object

### PR DESCRIPTION
**Pull request by aliuk2012**
### What problem does this pull request solve?

Create new QuestionForm object to replace using Page ActiveResource model directly. This means that the "Edit question" page just becomes another step in the add/edit question wizard. 

Updated the sequence diagrams https://github.com/alphagov/forms/pull/104

#### Current flow

```mermaid
erDiagram
    FORM{
        integer ID pk "this wouldn't necessarily be in sync with what is in forms_api"
    }
    session{
    
    }
    PAGES{
        integer ID pk
        integer forms_id fk
        integer user_id fk
        integer page_id "if page already exists"
        string question_text 
        string hint_text
        boolean is_optional
        string page_heading
        string guidance_markdown
        jsonb answer_settings
        datetime created_at
        datetime updated_at
    }
    PAGES_TYPE_OF_ANSWER_FORM{
        string answer_type
    
    }
    PAGES_ADDRESS_SETTINGS_FORM{
        boolean uk_address
        boolean international_address 
        string notes "input type becomes a hash instead of a string for this answer type eg. { input_type: { uk_address:, international_address: } }"
    }
    
    PAGES_DATE_SETTINGS_FORM{
        string input_type "can be 'date_of_birth' or 'other_date'"
        string notes "{ input_type:}"
    }

    PAGES_NAME_SETTINGS_FORM{
        string input_type "can be %w[full_name first_and_last_name first_middle_and_last_name]"
        boolean title_needed
        string notes "{ input_type:, title_needed: }"
    }

    PAGES_QUESTION_TEXT_FORM{
        string question_text "this is the same value that will be used for the final Page class"
        string notes "does not update answer_settings"
    }

    PAGES_SELECTIONS_SETTINGS_FORM{
        json seection_options "this is a list of hashes [{name: 'Option 1'}, {name: 'Option 2'}...]"
        boolean only_one_option "forms-runner uses this to control whether radio buttons or checkboxes are shown"
        boolean include_none_of_the_above "sets 'is_optional' and allows users to skip question"
        string notes "updates answer_settings and is_optional {is_option: include_none_of_the_above, answer_settings: { only_one_option:, selection_options: }}"
    }
    PAGES_TEXT_SETTINGS_FORM{
        string input_type "single_line | long_text]"
        string hint_text
        string notes "{ input_type: }"
    }

    PAGES_GUIDANCE_FORM{
        string page_heading
        string guidance_markdown
        string notes "Doesn't change answer_settings and only involves setting page page_heading and guidance_markdown"
    }
    FORM one to many PAGES :  has

    PAGES one to one session : reads
    PAGES one to one PAGES : "read/writes"
    session one to one PAGES_ADDRESS_SETTINGS_FORM : "read/writes"
    session one to one PAGES_TYPE_OF_ANSWER_FORM : "reads/writes"
    session one to one PAGES_DATE_SETTINGS_FORM: "reads/writes"
    session one to one PAGES_NAME_SETTINGS_FORM: "reads/writes"
    session one to one PAGES_QUESTION_TEXT_FORM: "reads/writes"
    session one to one PAGES_SELECTIONS_SETTINGS_FORM: "reads/writes"
    session one to one PAGES_TEXT_SETTINGS_FORM: "reads/writes"
    session one to one PAGES_GUIDANCE_FORM: "reads/writes"
    

```

#### Proposal

No need for Pages to read/write itself because the new `Pages::QuestionForm` (see far right on diagram) will read part of the data from session but ultimately forward its data on to Pages table for the time being. Once this is cleared up we can switch out sessions with our DraftQuestion ActiveRecord class.

```mermaid
erDiagram
    FORM{
        integer ID pk "this wouldn't necessarily be in sync with what is in forms_api"
    }
    session{
    
    }
    PAGES{
        integer ID pk
        integer forms_id fk
        integer user_id fk
        integer page_id "if page already exists"
        string question_text 
        string hint_text
        boolean is_optional
        string page_heading
        string guidance_markdown
        jsonb answer_settings
        datetime created_at
        datetime updated_at
    }
    PAGES_TYPE_OF_ANSWER_FORM{
        string answer_type
    
    }
    PAGES_ADDRESS_SETTINGS_FORM{
        boolean uk_address
        boolean international_address 
        string notes "input type becomes a hash instead of a string for this answer type eg. { input_type: { uk_address:, international_address: } }"
    }
    
    PAGES_DATE_SETTINGS_FORM{
        string input_type "can be 'date_of_birth' or 'other_date'"
        string notes "{ input_type:}"
    }

    PAGES_NAME_SETTINGS_FORM{
        string input_type "can be %w[full_name first_and_last_name first_middle_and_last_name]"
        boolean title_needed
        string notes "{ input_type:, title_needed: }"
    }

    PAGES_QUESTION_TEXT_FORM{
        string question_text "this is the same value that will be used for the final Page class"
        string notes "does not update answer_settings"
    }

    PAGES_SELECTIONS_SETTINGS_FORM{
        json seection_options "this is a list of hashes [{name: 'Option 1'}, {name: 'Option 2'}...]"
        boolean only_one_option "forms-runner uses this to control whether radio buttons or checkboxes are shown"
        boolean include_none_of_the_above "sets 'is_optional' and allows users to skip question"
        string notes "updates answer_settings and is_optional {is_option: include_none_of_the_above, answer_settings: { only_one_option:, selection_options: }}"
    }
    PAGES_TEXT_SETTINGS_FORM{
        string input_type "single_line | long_text]"
        string hint_text
        string notes "{ input_type: }"
    }

    PAGES_QUESTION_FORM{
        string question_text
        string hint_text
        boolean is_optional
        string answer_type
    }

    PAGES_GUIDANCE_FORM{
        string page_heading
        string guidance_markdown
        string notes "Doesn't change answer_settings and only involves setting page page_heading and guidance_markdown"
    }
    FORM one to many PAGES :  has

    PAGES one to one session : reads
    session one to one PAGES_ADDRESS_SETTINGS_FORM : "read/writes"
    session one to one PAGES_TYPE_OF_ANSWER_FORM : "reads/writes"
    session one to one PAGES_DATE_SETTINGS_FORM: "reads/writes"
    session one to one PAGES_NAME_SETTINGS_FORM: "reads/writes"
    session one to one PAGES_QUESTION_TEXT_FORM: "reads/writes"
    session one to one PAGES_SELECTIONS_SETTINGS_FORM: "reads/writes"
    session one to one PAGES_TEXT_SETTINGS_FORM: "reads/writes"
    session one to one PAGES_GUIDANCE_FORM: "reads/writes"
    session one to one PAGES_QUESTION_FORM : " reads/writes some data"
    

``` 

Trello card:  https://trello.com/c/Kw7ZwP16/1066-switch-edit-question-page-to-use-its-own-form-object

### Things to consider when reviewing

- Does it work when run on your machine?
- Is it clear what the code is doing?
- Do the commit messages explain why the changes were made?
- Are there all the unit tests needed?
- Has all relevant documentation been updated?

### CODE REVIEW

**Comment by lfdebrux** [comment]
> Create new QuestionForm object to replace using Page ActiveResource model directly. This means that the "Edit question" page just becomes another step in the add/edit question wizard.

Does this mean we're storing less in the session?

**Comment by aliuk2012** [comment]
@lfdebrux no we dont currently store some parts of the "Edit Question" page in session (eg. we dont store hint text). All we doing in this PR is making every page of the Add/Edit Question journey follow the Form Object pattern. What we currently have is the answer type, various answer settings pages, guidance all using Form Object pattern and just the "Edit question" page using Page ActiveResource model. 

This is all prep for us switching over from session to ActiveRecord. Once this PR is in  we can start switching each Form Object to write their data to DraftQuestion ActiveResource instead of session.

File: app/controllers/pages/questions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+  def create
+    answer_settings = session.dig(:page, :answer_settings)
+    page_heading = session.dig(:page, :page_heading)
+    guidance_markdown = session.dig(:page, :guidance_markdown)
+
+    @page = Page.new(page_params.merge(answer_settings:, page_heading:, guidance_markdown:))
+
+    @question_form = Pages::QuestionForm.new(page_params.merge(answer_settings:, page_heading:, guidance_markdown:))
+    if @question_form.valid? && @page.save
```
Comment (lfdebrux) [inline app/controllers/pages/questions_controller.rb:23]: Nice!

File: app/controllers/pages/questions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  end
+
+  def create
+    answer_settings = session.dig(:page, :answer_settings)
+    page_heading = session.dig(:page, :page_heading)
+    guidance_markdown = session.dig(:page, :guidance_markdown)
+
+    @page = Page.new(page_params.merge(answer_settings:, page_heading:, guidance_markdown:))
+
+    @question_form = Pages::QuestionForm.new(page_params.merge(answer_settings:, page_heading:, guidance_markdown:))
```
Comment (lfdebrux) [inline app/controllers/pages/questions_controller.rb:22]: Teensy tiny formatting suggestion, right now with the line breaks where they are I find it easy to miss the if statement, I think this would read better:

```suggestion
    @page = Page.new(page_params.merge(answer_settings:, page_heading:, guidance_markdown:))
    @question_form = Pages::QuestionForm.new(page_params.merge(answer_settings:, page_heading:, guidance_markdown:))

```

File: config/locales/en.yml [hunk 3]
```diff
@@ -10,6 +10,13 @@ en:
             question_text:
               blank: Enter a question
               too_long: Question text must be %{count} characters or less
```
Comment (lfdebrux) [inline config/locales/en.yml:12]: Do we still need these errors for the `page` model in here?

File: config/routes.rb [hunk 4]
```diff
@@ -98,10 +98,10 @@
           post "/address-settings" => "pages/address_settings#update", as: :address_settings_update
           get "/name-settings" => "pages/name_settings#edit", as: :name_settings_edit
           post "/name-settings" => "pages/name_settings#update", as: :name_settings_update
-          get "/" => "pages#edit", as: :edit_page
-          patch "/" => "pages#update", as: :update_page
           get "/guidance" => "pages/guidance#edit", as: :guidance_edit
           post "/guidance" => "pages/guidance#update", as: :guidance_update
+          get "/question" => "pages/questions#edit", as: :edit_question
+          post "/question" => "pages/questions#update", as: :update_question
```
Comment (lfdebrux) [inline config/routes.rb:104]: Why the change from `patch` to `post`?

**Review by lfdebrux — APPROVED** [review]
I think this all makes sense to me, and it seems to work on my machine. I have a couple of questions, but they're non-substantive so feel free to ignore them.

File: config/locales/en.yml (hunk 3 above)
Comment (aliuk2012) [reply config/locales/en.yml:12]: Good spot, I think we should keep them there for the time being. It should be tidied up once we move wholesale to ActiveRecord.

File: config/routes.rb (hunk 4 above)
Comment (aliuk2012) [reply config/routes.rb:104]: This is to bring it inline with all the other add/edit question step views. We are still making `patch` request to `forms-api`  this is only the request between the browser and forms-admin.

File: app/controllers/pages/questions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
+class Pages::QuestionsController < PagesController
```
[the remark this answers is not in the document]
Comment (aliuk2012) [reply app/controllers/pages/questions_controller.rb:1]: At the moment, its following what all the other add/edit pages controllers are doing but potentially it could inherit from a `BaseController` instead. Theres still more work to do on this but will keep this comment in mind as the refactoring goes along.

**Review by lfdebrux — APPROVED** [review]

File: app/views/pages/new.html.erb
[the remark this answers is not in the document]
Comment (aliuk2012) [reply app/views/pages/new.html.erb:1]: Yeah, we can but lets keep this as is for the time being. There are other views that we will eventually be creating too eg the "Check your question" page so we could do this later.
